components/blocks/cloud.js

```diff
@@ -1,45 +1,106 @@
-
-  if (height) {
-    CloudBlock = (
-      <section className={styles.Container}>
-        <iframe
-          loading="lazy"
-          src={`${src}`}
-          height={height}
-          className={styles.Iframe}
-          allow="camera;clipboard-read;clipboard-write;"
-          key={src}
-        />
-        {/* <a href={src} target="_blank" className={styles.Caption}>
-          (view standalone Streamlit app)
-        </a> */}
-      </section>
-    );
+// Arguments:
+//
+// - name: a string with the subdomain of the URL. That's the part between "https://" and
+//   ".streamlit.app". This is more future-proof that typing the full domain in case we change
+//   the URL scheme at some point.
+//
+// - path: an optional string with the app's subpath. This is the part after ".streamlit.app/"
+//   and before any query params.
+//
+// - query: an optional string with query params to add. Do not include the "?" at the front.
+//
+// - height: an optional string with a value CSS height. Must include a unit.
+//
+//
+// Arguments you shouldn't use by hand. Instead, these are meant to be used in table.js, which
+// needs to handle deprecated arguments we used to support in the iframes from our docstrings.
+//
+// - domain: a string with the app's full domain, like "foo.streamlit.app".
+//
+// - stylePlaceholder: a string with a regex placeholder like "$1".
+//
+//
+// For example:
+//
+//   <Cloud name="foo" query="embedOptions=show_padding" height="500px" />
+//   -> https://foo.streamlit.app/?embed=true&embed_options=show_padding
+//
+//   <Cloud name="foo" path="bar" query="embedOptions=show_padding&embedOptions=show_colored_line" />
+//   -> https://foo.streamlit.app/bar/?embed=true&embed_options=show_padding&embed_options=show_colored_line
+//
+const Cloud = ({ name, path, query, height, domain, stylePlaceholder }) => {
+  if (!domain) domain = `${name}.streamlit.app`;
+  if (domain.endsWith("/")) domain = domain.slice(0, -1);
+
+  if (path) {
+    if (!path.startsWith("/")) path = "/" + path;
   } else {
-    CloudBlock = (
-      <section className={styles.Container}>
-        <iframe
-          loading="lazy"
-          src={`${src}`}
-          className={classNames(styles.Iframe, styles.VideoAspectRatio)}
-          allow="camera;clipboard-read;clipboard-write;"
-          key={src}
-        />
-        {/* <a href={src} target="_blank" className={styles.Caption}>
-          (view standalone Streamlit app)
-        </a> */}
-      </section>
-    );
+    path = "";
   }
-  return CloudBlock;
+
+  const queryStr = query ? `&${query}` : "";
+
+  if (!height) height = "10rem";
+
+  const style = stylePlaceholder
+    ? // Hack so React lets us put a capture group placeholder in the style tag.
+      { ["--ignore-me"]: `42; ${stylePlaceholder}` }
+    : height
+      ? { height }
+      : null;
+
+  const frameSrc = `https://${domain}/~/+${path}?embed=true${queryStr}`;
+  const linkSrc = `https://${domain}${path}?utm_medium=oembed`;
```

Comment:
Hmm... that's a good point, but I don't have a good answer.

This code is used by two different code paths.
1. **Code that embeds frames via `<Cloud>`:** here we _could_ tell apart embed-related query params (which should be excluded from `linkSrc`) from normal query params (which should be included in `linkSrc`).
2. **Code in `table.js` that embeds frames from docstrings:** this code cannot tell those apart, since it uses a Regex. Anything else would be a lot more complex.

By the way, this is a non-issue today because `<Cloud>`'s `query` argument is only used in two places, both of which use it to set embed-related params.

**So perhaps the simpler solution here is: just rename the `query` argument to `embedParams`**, and don't include those in `linkSrc`.

And if you _really_ want to have the ability to add random query params later, I can add another argument called `queryParams` and make sure they're included in `linkSrc`.

Reply:
Ok, I have a different solution.

`<Cloud query={...} >` can receive any sort of query param. If they're embed params, we don't include in the "Fullscreen" button. But all others we do. So this works as you wanted.

However, if you set query params in the `.. output::` directive (like `https://foo.streamlit.app/?param=123`) those will _never_ be treated as embed params. So they will _always_ appear in the "Fullscreen" link.

I'm committing that solution to this PR right now. LMKWYT!

Reply:
Yeah, I figured we'd either parse `query` and drop `embed_options` if present, or create an `embed_options` parameter. It's not currently a mainstay, so the former is perfectly fine.